### PR TITLE
Added optional argument `batch_first` to the transformer module

### PR DESCRIPTION
**Pull request by scott0123**
Fixes an open issue in #32590 while maintaining BC.

### CODE REVIEW

**Comment by scott0123** [comment]
@zhangguanheng66 Can you review this?

**Comment by zhangguanheng66** [comment]
Not sure if we really want to add `batch_first` argument. IMO, this issue doesn't really bother users that much.

**Review by zhangguanheng66 — COMMENTED** [review]
Not sure if we really want to add `batch_first` argument. IMO, this issue doesn't really bother users that much.

**Comment by scott0123** [comment]
It is an optional argument that makes it more consistent and in line with our RNN modules which all have this as an optional argument.

This was requested in #25100 #25132, and this forum post: https://discuss.pytorch.org/t/multiheadattention-transformer-with-batch-first/54171
Your post in #32590 also has this as a potential TODO item.

Let me know what you think. I've also updated the test cases in `test_nn.py` to include this argument.

File: test/test_nn.py [hunk 1]
```diff
@@ -5303,11 +5305,11 @@ def update_shape(shape, dim, new_dim_size):
         # Correct activations
         for activation in activations:
             model = getattr(nn, model_name)(d_model, nhead, num_encoder_layers, num_decoder_layers,
-                                            dim_feedforward, dropout, activation)
+                                            dim_feedforward, batch_first, dropout, activation)
         # Incorrect activation
         with self.assertRaises(RuntimeError):
             model = getattr(nn, model_name)(d_model, nhead, num_encoder_layers, num_decoder_layers,
-                                            dim_feedforward, dropout, wrong_activation)
+                                            dim_feedforward, batch_first, dropout, wrong_activation)
```
Comment (zhangguanheng66) [inline test/test_nn.py:5312]: I think if you add batch_first to the end, you won't need to add it between other arguments.

File: test/test_nn.py [hunk 2]
```diff
@@ -4571,21 +4571,22 @@ def test_pack_sequence_batch_sizes_throw(self):
             b = torch.tensor([1, 1, 1, 1, 1], device='cuda')
             input = nn.utils.rnn.PackedSequence(a, b)
 
-    def test_Transformer_cell(self):
+    def test_transformer_cell(self):
```
Comment (zhangguanheng66) [inline test/test_nn.py:4574]: you need to add a test to cover the batch_first argument with `True`

File: test/test_nn.py (hunk 1 above)
Comment (scott0123) [reply test/test_nn.py:5312]: This is true, and I initially thought about doing that. However, `batch_first` has traditionally been an argument before `dropout` (see RNN docs), so that is my rational for putting it there.

File: test/test_nn.py (hunk 2 above)
Comment (scott0123) [reply test/test_nn.py:4573]: Done in commit 248d44b

Comment (scott0123) [reply test/test_nn.py:4574]: Thanks, writing the test case helped uncover some potential problems. Now this has all been resolved as of commit 8bea5f0.

File: test/test_nn.py [hunk 3]
```diff
@@ -4604,6 +4605,67 @@ def test_Transformer_cell(self):
...
+        tgt_batch_first = tgt.permute(1, 0, 2)
+        output_batch_first = transformer_batch_first(src_batch_first, tgt_batch_first,
+                                                     src_mask=src_mask,
+                                                     tgt_mask=tgt_mask,
+                                                     memory_mask=memory_mask,
+                                                     src_key_padding_mask=src_key_padding_mask,
+                                                     tgt_key_padding_mask=tgt_key_padding_mask,
+                                                     memory_key_padding_mask=memory_key_padding_mask)
+        output_ref = output_batch_first.permute(1, 0, 2)
+        np.testing.assert_allclose(output.detach().numpy(), output_ref.detach().numpy(), atol=1e-5)
```
Comment (zhangguanheng66) [inline test/test_nn.py:4667]: You will need a numpy decorator if your test depends on numpy. I'd avoid the numpy dependency and `torch.testing.assert_allclose` (though some of transformer tests did use numpy)

File: torch/nn/modules/transformer.py [hunk 4]
```diff
@@ -177,15 +184,21 @@ def forward(self, src, mask=None, src_key_padding_mask=None):
...
+            output = src
 
         for mod in self.layers:
             output = mod(output, src_mask=mask, src_key_padding_mask=src_key_padding_mask)
 
         if self.norm is not None:
             output = self.norm(output)
 
-        return output
+        if self.batch_first:
```
Comment (zhangguanheng66) [inline torch/nn/modules/transformer.py:198]: The reason that we have to permute within `TransformerEncoder` and `TransformerDecoder` is that the encoder/decoder layers don't support `batch_first`. Should we consider to add `batch_first` support to them?

File: torch/nn/modules/transformer.py [hunk 5]
```diff
@@ -159,11 +165,12 @@ class TransformerEncoder(Module):
     """
     __constants__ = ['norm']
 
-    def __init__(self, encoder_layer, num_layers, norm=None):
+    def __init__(self, encoder_layer, num_layers, norm=None, batch_first=False):
```
Comment (zhangguanheng66) [inline torch/nn/modules/transformer.py:168]: Users may pass a custom encoder/decoder layers here with batch first or not. We need to be careful in that case.

Comment (scott0123) [reply torch/nn/modules/transformer.py:168]: What's the best way to handle this? If we add it as an optional argument to the custom encoder/decoder, it may break BC depending on the user's custom encoder/decoder implementation.

File: torch/nn/modules/transformer.py (hunk 4 above)
Comment (scott0123) [reply torch/nn/modules/transformer.py:198]: This is a good question. It would certainly be less overhead if the encoder/decoder layers themselves have an implementation that natively supports `batch_first` or not. I guess it would be worth figuring out if this permute operation is taking up a significant fraction of the time per forward/backward computation (e.g. >1%)

File: test/test_nn.py (hunk 3 above)
Comment (scott0123) [reply test/test_nn.py:4667]: Yes. I might go ahead and change this and maybe also change the other transformer tests to all use `torch.testing.assert_allclose` too (if applicable).

**Comment by scott0123** [comment]
Any idea why the FB internal test suite is failing? My best guess is that in their custom test cases they did not use the preferred optional argument passing scheme when instantiating the `Transformer` class (AKA ultimately an issue with argument ordering)

File: torch/nn/modules/transformer.py (hunk 5 above)
Comment (scott0123) [reply torch/nn/modules/transformer.py:169]: One way would be to use the `inspect` module's `inspect.getfullargspec` to check if the custom encoder/decoder has the `batch_first` argument, and only pass that argument along to the custom encoder/decoder if it does. This feels inelegant to me, though.

**Comment by scott0123** [comment]
@zhangguanheng66 Possible to merge this now, or how do we move forward?

**Comment by zhangguanheng66** [comment]
> @zhangguanheng66 Possible to merge this now, or how do we move forward?

I'd not to rush and merge it since the transformer PR might affect many users. I will have a close look in the next few days.

File: test/test_nn.py [hunk 6]
```diff
@@ -4707,18 +4707,18 @@ def test_transformerencoderlayer(self):
         ref_output = torch.Tensor([[[2.272644, 0.119035, -0.691669, 0.153486]],
                                    [[2.272644, 0.119035, -0.691669, 0.153486]]])
         self.assertEqual(tuple(result.shape), tuple(ref_output.shape))
-        torch.testing.assert_allclose(result, ref_output)
+        torch.testing.assert_allclose(result, ref_output, rtol=1e-2, atol=1e-5)
```
Comment (zhangguanheng66) [inline test/test_nn.py:4710]: I'm concerned to loose the failure criteria because the original passes the tests.

**Comment by zhangguanheng66** [comment]
@malmaud Could you take an additional review and give some feedback?

File: test/test_nn.py (hunk 6 above)
Comment (scott0123) [reply test/test_nn.py:4710]: The current `torch.testing.assert_allclose` function asserts that both `rtol` (relative tolerance) and `atol` (absolute tolerance) must be specified. Previously in the numpy version of this test case,  only `atol` was set to `1e-5`.
My changes ensure that this is still maintained, while adding an extra `rtol=1e-2`. This passes the present test case and should not loosen the criteria too much in my opinion.

**Comment by malmaud** [comment]
I don't think I know this part of the codebase well enough to comment on the PR itself, but it does accomplish what I was hoping for, so thank you @scott0123.

**Comment by davidkartchner** [comment]
@zhangguanheng66  Is there a plan to merge this at some point?  I would find this feature very useful, especially for comparing current transformer-based architectures with older RNN ones without needing to rewrite dataloading scripts.

**Comment by zhangguanheng66** [comment]
> @zhangguanheng66 Is there a plan to merge this at some point? I would find this feature very useful, especially for comparing current transformer-based architectures with older RNN ones without needing to rewrite dataloading scripts.

I think you can update `collate_fn` in `dataloader` to achieve this goal.

**Comment by Kartikinator** [comment]
@zhangguanheng66 Hi, can this be merged anytime soon? I would really appreciate this feature as it would line up with the way I pass my data through other RNNs.

**Review by zhangguanheng66 — CHANGES_REQUESTED** [review]
The PR looks good to me. Could you @scott0123 rebase with the master branch to resolve the conflict?

**Comment by scaomath** [comment]
> The PR looks good to me. Could you @scott0123 rebase with the master branch to resolve the conflict?

Oh, I did not realize there was already a PR on this issue. @zhangguanheng66 could you please check the PR #55285 ? The base I worked on should be most up-to-date one, but essentially did what @scott0123 did while adding the `contiguous()` part. On my local end I use this for mesh-based application with a custom attention, so there are multiple dims not just 3, I think using `transpose` should be better than `permute` to be more future proof.

**Review by zhangguanheng66 — COMMENTED** [review]
Left a similar comment in https://github.com/pytorch/pytorch/pull/55285. Do we want to add transpose/permute on the transformer level or the MHA level?

cc @jbschlosser

**Comment by zhangguanheng66** [comment]
> > The PR looks good to me. Could you @scott0123 rebase with the master branch to resolve the conflict?
> 
> Oh, I did not realize there was already a PR on this issue. @zhangguanheng66 could you please check the PR #55285 ? The base I worked on should be most up-to-date one, but essentially did what @scott0123 did while adding the `contiguous()` part. On my local end I use this for mesh-based application with a custom attention, so there are multiple dims not just 3, I think using `transpose` should be better than `permute` to be more future proof.

Exactly. Like [MultiheadAttentionContainer](https://github.com/pytorch/text/blob/master/torchtext/nn/modules/multiheadattention.py#L5), we would like to support broadcast with the inputs more than 3 dim.
